pkg/querier/queryrange/downstreamer.go

```diff
@@ -245,3 +254,372 @@ func ResponseToResult(resp queryrangebase.Response) (logqlmodel.Result, error) {
+	// {'a': 1, 'b': 0}
+	// [b, a]
+}
+
+// first order by timestamp, then by labels
+func (acc *accumulatedStreams) Less(i, j int) bool {
+	// order by the 'oldest' entry in the stream
+	if a, b := acc.streams[i].Entries[len(acc.streams[i].Entries)-1].Timestamp, acc.streams[j].Entries[len(acc.streams[j].Entries)-1].Timestamp; !a.Equal(b) {
+		return acc.less(a, b)
+	}
+	return acc.streams[i].Labels <= acc.streams[j].Labels
+}
+
+func (acc *accumulatedStreams) less(a, b time.Time) bool {
+	// use after for stable sort
+	if acc.order == logproto.FORWARD {
+		return !a.After(b)
+	}
+	return !b.After(a)
+}
+
+func (acc *accumulatedStreams) Push(x any) {
+	s := x.(*logproto.Stream)
+	if len(s.Entries) == 0 {
+		return
+	}
+
+	if room := acc.limit - acc.count; room >= len(s.Entries) {
+		if i, ok := acc.Find(s.Labels); ok {
+			// stream already exists, append entries
+
+			// these are already guaranteed to be sorted
+			// Reasoning: we shard subrequests so each stream exists on only one
+			// shard. Therefore, the only time a stream should already exist
+			// is in successive splits, which are already guaranteed to be ordered
+			// and we can just append.
+			acc.appendTo(acc.streams[i], s)
+
+			return
+		}
+
+		// new stream
+		acc.addStream(s)
+		return
+	}
+
+	// there's not enough room for all the entries,
+	// so we need to
+	acc.push(s)
+}
+
+// there's not enough room for all the entries.
+// since we store them in a reverse heap relative to what we _want_
+// (i.e. the max value for FORWARD, the min value for BACKWARD),
+// we test if the new entry is better than the worst entry,
+// swapping them if so.
+func (acc *accumulatedStreams) push(s *logproto.Stream) {
+	worst, ok := acc.top()
+	room := min(acc.limit-acc.count, len(s.Entries))
+
+	if !ok {
+		if room == 0 {
+			// special case: limit must be zero since there's no room and no worst entry
+			return
+		}
+		s.Entries = s.Entries[:room]
+		// special case: there are no entries in the heap. Push entries up to the limit
+		acc.addStream(s)
+		return
+	}
+
+	// since entries are sorted by timestamp from best -> worst,
+	// we can discard the entire stream if the incoming best entry
+	// is worse than the worst entry in the heap.
+	cutoff := sort.Search(len(s.Entries), func(i int) bool {
+		// TODO(refactor label comparison -- should be in another fn)
+		if worst.Equal(s.Entries[i].Timestamp) {
+			return acc.streams[0].Labels < s.Labels
+		}
+		return acc.less(s.Entries[i].Timestamp, worst)
+	})
+	s.Entries = s.Entries[:cutoff]
+
+	for i := 0; i < len(s.Entries) && acc.less(worst, s.Entries[i].Timestamp); i++ {
+
+		// push one entry at a time
+		room = acc.limit - acc.count
+		// pop if there's no room to make the heap small enough for an append;
+		// in the short path of Push() we know that there's room for at least one entry
+		if room == 0 {
+			acc.Pop()
+		}
+
+		cpy := *s
+		cpy.Entries = []logproto.Entry{s.Entries[i]}
+		acc.Push(&cpy)
```

Comment:
Instead of pushing each entry individually, could we batch entries and push them at once?

Reply:
We need to compare them individually because we may only need 1, or 2, or n entries from the incoming stream and we can only see the "worst" value on the heap. That means we're only able to detect "is the best incoming value better than the current worst value on the heap". This is why each time that returns `true`, we `Pop()` the oldest worst value and `Push()` the new value and reconfigure the stream orders within the heap on both operations. There's probably some optimizations that could be made around tracking that in the future, but I don't think it's worth the current complexity.

Reply:
Got it!